File: content-planner-motia/steps/scrape.step.py
from pydantic import BaseModel, HttpUrl
from typing import Dict, Any
import requests
import sys
import os

# Add the parent directory to the path to import config
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from config.index import config as app_config

class ScrapeInput(BaseModel):
    requestId: str
    url: HttpUrl
    timestamp: int

config = {
    'type': 'event',
    'name': 'ScrapeArticle',
    'description': 'Scrapes article content using Firecrawl',
    'subscribes': ['scrape-article'],
    'emits': ['analyze-content'],
    'input': ScrapeInput,
    'flows': ['content-generation']
}

async def handler(input: Dict[str, Any], context):
    logger = context.logger
    state = context.state
    emit = context.emit
    trace_id = context.traceId
    
    url = input['url']
    request_id = input['requestId']
    timestamp = input['timestamp']
    
    logger.info(f"🕷️ Scraping article: {url}")
    
    # Use Firecrawl API directly via HTTP requests
    firecrawl_url = "https://api.firecrawl.dev/v1/scrape"
    headers = {
        'Authorization': f'Bearer {app_config.firecrawl["api_key"]}',
        'Content-Type': 'application/json'
    }
    
    payload = {
        'url': str(url),
        'formats': ['markdown'],
        'onlyMainContent': True
    }
    
    try:
        response = requests.post(firecrawl_url, json=payload, headers=headers)
        response.raise_for_status()
        scrape_result = response.json()
        
        if not scrape_result.get('success'):
            raise Exception(f"Firecrawl scraping failed: {scrape_result.get('error', 'Unknown error')}")
        
        content = scrape_result.get('data', {}).get('markdown', '')
        title = scrape_result.get('data', {}).get('metadata', {}).get('title', 'Untitled Article')
        
        logger.info(f"✅ Successfully scraped: {title} ({len(content) if content else 0} characters)")
        
        await emit({
            'topic': 'analyze-content',
            'data': {
                'requestId': request_id,
                'url': str(url),
                'title': title,
                'content': content,
                'timestamp': timestamp
            }
        })
        
    except requests.exceptions.RequestException as e:
        logger.error(f"HTTP request failed: {str(e)}")
        raise Exception(f"Firecrawl scraping failed: {str(e)}")
    except Exception as e:
        logger.error(f"Scraping failed: {str(e)}")
        raise

File: content-planner-motia/steps/scrape.step.ts
import { EventConfig, Handlers } from 'motia'
import { z } from 'zod'
import FirecrawlApp from '@mendable/firecrawl-js'

const config = require('../config')

export const config: EventConfig = {
  type: 'event',
  name: 'ScrapeArticle',
  description: 'Scrapes article content using Firecrawl',
  subscribes: ['scrape-article'],
  emits: ['analyze-content'],
  input: z.object({
    requestId: z.string(),
    url: z.string().url(),
    timestamp: z.number()
  }),
  flows: ['content-generation']
}

export const handler: Handlers['ScrapeArticle'] = async (input, { emit, logger }) => {
  logger.info(`🕷️ Scraping article: ${input.url}`)

  const app = new FirecrawlApp({ apiKey: config.firecrawl.apiKey })

  const scrapeResult = await app.scrapeUrl(input.url, {
    formats: ['markdown'],
    onlyMainContent: true
  })

  if (!scrapeResult.success) {
    throw new Error(`Firecrawl scraping failed: ${scrapeResult.error}`)
  }

  const content = scrapeResult.markdown  
  const title = scrapeResult.metadata?.title || 'Untitled Article'

  logger.info(`✅ Successfully scraped: ${title} (${content?.length ?? 0} characters)`)

  await emit({
    topic: 'analyze-content',
    data: {
      requestId: input.requestId,
      url: input.url,
      title,
      content,
      timestamp: input.timestamp
    }
  })
}